### PR TITLE
Initial  commit of zookeeper example and statefulsets dir

### PR DESCRIPTION
**Pull request by kow3ns**

### CODE REVIEW

**Comment by kow3ns** [comment]
@erictune @enisoc @foxish

**Comment by enisoc** [comment]
Reviewed 7 of 10 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[statefulsets/zookeeper/Dockerfile, line 48 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZnYwlEAu0GxNbjEUA:-KYZnYwlEAu0GxNbjEUB:b-17kjv3) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/Dockerfile#L48)):*
> ```Dockerfile
> RUN set -x \
> 	&& useradd $ZK_USER \
>     && id $ZK_USER \
> ```

Since you rely on the uid/gid being specific numeric values in the k8s manifest, you can "assert" them here instead of just printing:

```
    && [ `id -u $ZK_USER` -eq 1000 ] \
    && [ `id -g $ZK_USER` -eq 1000 ] \
```

Then the docker build will fail if the assumption is violated because of a change in an upstream image or tool.

---

*[statefulsets/zookeeper/README.md, line 54 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZotou_sWkg7KBK-x3:-KYZotou_sWkg7KBK-x4:be61kp8) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L54)):*
> ```Markdown
> The Stateful Set configuration must match the Headless Service, and it must provide the number of 
> replicas. In the example below we request a ZooKeeper ensemble of size 3. 
> **As weighed quorums are not supported, it is imperative that an odd number of replicas be chosen.
> ```

Do you mean "weighted"?

---

*[statefulsets/zookeeper/README.md, line 77 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZpDcnUtMow0IGFtmc:-KYZpDcnUtMow0IGFtmd:b-hm3ch1) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L77)):*
> ```Markdown
>       - name: k8szk
>         imagePullPolicy: Always
>         image: kowens/k8szk
> ```

Don't forget to update this copy of the yaml to the gcr.io image.

---

*[statefulsets/zookeeper/README.md, line 140 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZpWNdEcyYrwiCcObI:-KYZpWNdEcyYrwiCcObJ:b595bx3) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L140)):*
> ```Markdown
> outdated logs and snapshots.
> 
> If you do not have a existing retention policy and backup procedure, and if you are comfortable with 
> ```

"an* existing"

---

*[statefulsets/zookeeper/README.md, line 193 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZpl6glbr4zs9YiuRm:-KYZpl6glbr4zs9YiuRn:bb6wegf) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L193)):*
> ```Markdown
> Currently, the use of Persistent Volumes to provide durable, network attached storage is mandatory.
> **If you use the provided image with emptyDirs, you will likely suffer a data loss.** The example 
> below demonstrates how to request a dynamically provisioned persistent volume of 20 Gib.
> ```

GiB*

---

*[statefulsets/zookeeper/zookeeper.yaml, line 58 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZrHTHfpUuwku-9oQT:-KYZrHTHfpUuwku-9oQU:b-2s65x2) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/zookeeper.yaml#L58)):*
> ```YAML
>             {
>               "podAntiAffinity": {
>                 "requiredDuringSchedulingRequiredDuringExecution": [{
> ```

As mentioned in https://github.com/kubernetes/kubernetes.github.io/pull/1894#issuecomment-266081871 I'm wondering if *preferred* might be better. Whatever we decide there will probably apply here too. I'm just leaving a link here so they'll be consistent.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134)*

**Comment by foxish** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[statefulsets/zookeeper/README.md, line 7 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY_OaMq-K8lv2Mf4nZp:-KY_OaMrlLHFWwEvwHyF:blnn5tk) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L7)):*
> ```Markdown
> ##Limitations
> 1. Scaling is not currently supported. Once an ensemble is created, it is not safe to patch the 
> replicas of the Stateful Set. This is primarily due to the lack of the ability to update ensemble's 
> ```

This sentence could be simplified.

---

*[statefulsets/zookeeper/README.md, line 9 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY_OgMzF4UBYyMfGdeO:-KY_OgMzF4UBYyMfGdeP:b-k5d8lx) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L9)):*
> ```Markdown
> replicas of the Stateful Set. This is primarily due to the lack of the ability to update ensemble's 
> membership in a safe way in ZooKeeper 3.4.9 (The current stable release). When ZooKeeper 3.5.x 
> advances from alpha to stable we will consider adding this feature.
> ```

I think making statements about the future was something we were avoiding elsewhere in documentation. Should be avoided here as well? 

---

*[statefulsets/zookeeper/README.md, line 133 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY__7eFMbCXS21p368r:-KY__7eFMbCXS21p368s:b-qas1ij) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L133)):*
> ```Markdown
> 
> ####Data Retention Configuration
> **ZooKeeper does not, by default, purge old transactions logs or snapshots. No, never. Yes, this can cause 
> ```

nit: I found the "No" and the "Yes" a little hard to parse in a quick read.

---

*[statefulsets/zookeeper/README.md, line 187 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY_aHCUP_IGAHIulRbA:-KY_aHCVcb9BuwTcx0lz:b-lgafho) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L187)):*
> ```Markdown
> ```yaml
>   volumeMounts:
>           - name: datadir
> ```

Indentation seems a bit off.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134)*

**Comment by kow3ns** [comment]
Review status: 8 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[statefulsets/zookeeper/Dockerfile, line 48 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZnYwlEAu0GxNbjEUA:-KYv0u75YpfFnECeQ56w:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/Dockerfile#L48)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

Since you rely on the uid/gid being specific numeric values in the k8s manifest, you can "assert" them here instead of just printing:

```
    && [ `id -u $ZK_USER` -eq 1000 ] \
    && [ `id -g $ZK_USER` -eq 1000 ] \
```

Then the docker build will fail if the assumption is violated because of a change in an upstream image or tool.
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 7 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY_OaMq-K8lv2Mf4nZp:-KYv2hbvj9QfMXbFTffK:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L7)):*
<details><summary><i>Previously, foxish (Anirudh Ramanathan) wrote…</i></summary><blockquote>

This sentence could be simplified.
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 9 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY_OgMzF4UBYyMfGdeO:-KYv10R0bET2hZI2p8uH:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L9)):*
<details><summary><i>Previously, foxish (Anirudh Ramanathan) wrote…</i></summary><blockquote>

I think making statements about the future was something we were avoiding elsewhere in documentation. Should be avoided here as well? 
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 54 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZotou_sWkg7KBK-x3:-KYv1DQoyma6ZtBn-lmf:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L54)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

Do you mean "weighted"?
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 77 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZpDcnUtMow0IGFtmc:-KYv1Fskwqgyu5mCSWjp:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L77)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

Don't forget to update this copy of the yaml to the gcr.io image.
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 133 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY__7eFMbCXS21p368r:-KYv1fR4mDS70t6Xte-2:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L133)):*
<details><summary><i>Previously, foxish (Anirudh Ramanathan) wrote…</i></summary><blockquote>

nit: I found the "No" and the "Yes" a little hard to parse in a quick read.
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 140 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZpWNdEcyYrwiCcObI:-KYv1i3jnTCXsv_F4XT4:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L140)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

"an* existing"
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 187 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY_aHCUP_IGAHIulRbA:-KYv2GiAySQMYN8H-szq:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L187)):*
<details><summary><i>Previously, foxish (Anirudh Ramanathan) wrote…</i></summary><blockquote>

Indentation seems a bit off.
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 193 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZpl6glbr4zs9YiuRm:-KYv289-1Y-kK1dvkZjo:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L193)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

GiB*
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/zookeeper.yaml, line 58 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KYZrHTHfpUuwku-9oQT:-KYv1nlmP50wh7hEXujc:beto4oy) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/zookeeper.yaml#L58)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

As mentioned in https://github.com/kubernetes/kubernetes.github.io/pull/1894#issuecomment-266081871 I'm wondering if *preferred* might be better. Whatever we decide there will probably apply here too. I'm just leaving a link here so they'll be consistent.
</blockquote></details>

We are going with hard spreading for the blog, tutorial, and container. At least for now.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134)*

**Comment by kow3ns** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134)*

**Comment by enisoc** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-:-KYyEC5NtkGJDP4IEfaH:bnfp4nl)*

**Comment by kow3ns** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[statefulsets/zookeeper/README.md, line 7 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY_OaMq-K8lv2Mf4nZp:-KYyVZE48b3ri787RfFL:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L7)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[statefulsets/zookeeper/README.md, line 9 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134#-KY_OgMzF4UBYyMfGdeO:-KYyVZp0NND4Picvh26w:b-896fix) ([raw file](https://github.com/kubernetes/contrib/blob/11bb6421926bb1afb6ce86aa5cf78e417e4cdb7c/statefulsets/zookeeper/README.md#L9)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/2134)*
